internal/rpc/httpproxy.go

```diff
@@ -7,98 +7,117 @@ import (
-// It tries for a controller, if it errors, it logs the error and go to the next, if no controller responds it returns a 504.
-func ProxyHTTP(ctx context.Context, ctl *dbmodel.Controller, w http.ResponseWriter, req *http.Request) error {
+func ProxyHTTP(ctx context.Context, ctl ControllerProxy, w http.ResponseWriter, req *http.Request) {
+	urls, err := getControllerAddresses(ctl.Controller)
+	if err != nil {
+		zapctx.Error(ctx, "failed to get controller addresses", zap.Error(err))
+		http.Error(w, fmt.Sprintf("failed to get controller addresses: %v", err), http.StatusInternalServerError)
+		return
+	}
+
 	var tlsConfig *tls.Config
-	if ctl.CACertificate != "" {
+	if ctl.Controller.CACertificate != "" {
 		cp := x509.NewCertPool()
-		ok := cp.AppendCertsFromPEM([]byte(ctl.CACertificate))
+		ok := cp.AppendCertsFromPEM([]byte(ctl.Controller.CACertificate))
 		if !ok {
 			zapctx.Warn(ctx, "no CA certificates added")
 		}
 		tlsConfig = &tls.Config{
 			RootCAs:    cp,
-			ServerName: ctl.TLSHostname,
+			ServerName: ctl.Controller.TLSHostname,
 			MinVersion: tls.VersionTLS12,
 		}
 	}
 
-	if ctl.PublicAddress != "" {
-		err := doRequest(ctx, w, req, httpOptions{
-			TLSConfig: tlsConfig,
-			URL:       createURLWithNewHost(*req.URL, ctl.PublicAddress),
-		})
-		if err == nil {
-			return nil
-		}
+	// transport is the default HTTP transport config with TLS configuration.
+	transport := &http.Transport{
+		TLSClientConfig:       tlsConfig,
+		Proxy:                 http.ProxyFromEnvironment,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
 	}
-	for _, hps := range ctl.Addresses {
-		for _, hp := range hps {
-			err := doRequest(ctx, w, req, httpOptions{
-				TLSConfig: tlsConfig,
-				URL:       createURLWithNewHost(*req.URL, fmt.Sprintf("%s:%d", hp.Value, hp.Port)),
-			})
-			if err == nil {
-				return nil
-			} else {
-				zapctx.Error(ctx, "failed to proxy request: continue to next addr", zaputil.Error(err))
-			}
-		}
+
+	if len(urls) == 0 {
+		zapctx.Error(ctx, "no controller addresses found", zap.String("controller", ctl.Controller.Name))
+		http.Error(w, "no controller addresses found", http.StatusInternalServerError)
+		return
 	}
 
-	return errgo.New("couldn't reach a valid address for controller")
-}
+	if len(urls) > 1 {
+		rand.Shuffle(len(urls), func(i, j int) {
+			urls[i], urls[j] = urls[j], urls[i]
+		})
+	}
 
-func doRequest(ctx context.Context, w http.ResponseWriter, req *http.Request, opt httpOptions) error {
-	client := &http.Client{
-		Transport: &http.Transport{
-			TLSClientConfig: opt.TLSConfig,
+	// TODO: Consider implementing a better load balancing mechanism that handles
+	// multiples URLs and handles failing backends gracefully e.g. try send to first
+	// URL and on failure, try second, etc.
+	proxy := &httputil.ReverseProxy{
```

Comment:
It needs to use our zap logger, see:
```go
    // ErrorLog specifies an optional logger for errors
    // that occur when attempting to proxy the request.
    // If nil, logging is done via the log package's standard logger.
    ErrorLog *log.Logger
```

Reply:
Also, do any of these endpoints need instant streaming? Because if we ever proxy one where user experience is important. We'd need flush interval to be set to a negative value I think? Probably not a cause for concern right now but something to be aware of?